plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPropertyPage.java

```diff
@@ -105,109 +116,92 @@ public void run() {
-      return new StandardDeployPreferencesPanel(
-          container, project, loginService, getLayoutChangedHandler(), false /* requireValues */,
-          new ProjectRepository(googleApiFactory));
-    } else if (AppEngineFlexFacet.hasAppEngineFacet(facetedProject)) {
-      setTitle(Messages.getString("flex.page.title"));
-      return new FlexDeployPreferencesPanel(container, project);
+    if (facetedProject != null && AppEngineStandardFacet.hasAppEngineFacet(facetedProject)) {
+      createStandardPanelIfNeeded();
+      stackLayout.topControl = standardPreferencesPanel;
+      databindingSupport =
+          PreferencePageSupport.create(this, standardPreferencesPanel.getDataBindingContext());
+    } else if (facetedProject != null && AppEngineFlexFacet.hasAppEngineFacet(facetedProject)) {
+      createFlexPanelIfNeeded();
+      stackLayout.topControl = flexPreferencesPanel;
+      databindingSupport =
+          PreferencePageSupport.create(this, flexPreferencesPanel.getDataBindingContext());
     } else {
-      logger.log(Level.WARNING,
-          "App Engine Deployment property page is only visible if project contains an App Engine facet");
-      return null;
+      createBlankPanelIfNeeded();
+      stackLayout.topControl = blankPreferencesPanel;
+      databindingSupport =
+          PreferencePageSupport.create(this, blankPreferencesPanel.getDataBindingContext());
     }
-  }
+    container.layout();
 
-  @Override
-  public void setVisible(boolean visible) {
-    super.setVisible(visible);
-    evaluateFacetConfiguration();
   }
 
-  /**
-   * Checks to see if the project associated with this Property dialog still has a valid
-   * App Engine facet. If it does, allow messages to be displayed. If it doesn't display
-   * appropriate error message and prevent other messages from being displayed.
-   */
-  private void evaluateFacetConfiguration() {
-    if (isStandardPanel && !AppEngineStandardFacet.hasAppEngineFacet(facetedProject)) {
-      IProjectFacet projectFacet = ProjectFacetsManager.getProjectFacet(AppEngineStandardFacet.ID);
-      invalidFacetConfigErrorMessage =
-          Messages.getString("invalid.deploy.page.state", projectFacet.getLabel());
-      updatePageBasedOnFacetConfig(false);
-    } else if (!isStandardPanel && !AppEngineFlexFacet.hasAppEngineFacet(facetedProject)) {
-      IProjectFacet projectFacet = ProjectFacetsManager.getProjectFacet(AppEngineFlexFacet.ID);
-      invalidFacetConfigErrorMessage =
-          Messages.getString("invalid.deploy.page.state", projectFacet.getLabel());
-      updatePageBasedOnFacetConfig(false);
-    } else {
-      invalidFacetConfigErrorMessage = "";
-      updatePageBasedOnFacetConfig(true);
+  private void createBlankPanelIfNeeded() {
+    if (blankPreferencesPanel == null) {
+      blankPreferencesPanel = new BlankDeployPreferencesPanel(container);
+      blankPreferencesPanel.setFacetPageSelector(new Runnable() {
+        @Override
+        public void run() {
```

Comment:
The reason I've put it here is that the panel is not aware of the `PreferenceDialog`, so it would take some `getParent()` and `instanceof` calls to find where it is.